### ISSUE TITLE
Enter/Exit animations

### ISSUE THREAD

**Issue by joshwcomeau** [joshwcomeau/react-flip-move#18]
A few people have asked about Enter/Exit animations. 

I think it's possible to keep in line with the [FLIP](https://aerotwist.com/blog/flip-your-animations/) methodology. Here's how I envision exit animations working:

* When the component receives new props, we check to see if any items are about to be removed
* We instead keep those elements, but apply an immediate style (something like `transform: scale(0, 0);`). The idea is to make the element take up 0px worth of space, so that when the layout is recalculated, its siblings treat it as though it wasn't there.
* We then do the FLIP animation. There's some flexibility here, we can have presets for _how_ removed items should be transitioned. They can fade out and shrink, or even just constrict in one axis (a vertical list could have removed items flattened!). I'm excited by this; it allows for a lot of possible customization.
* Once the animation is complete, we remove those nodes from the DOM by re-rendering directly from the props provided.

The same process could work for Enter animations. We'd start it at `transform: scale(0,0);`, and expand it to its normal size while other elements shift out of the way.

This is a non-trivial change to the source code. We'd need to find a way to "hold on" to the items that are being removed. There may be unforeseen complications.

I'll try and dedicate some time in the next month to test this idea out. If anyone wants to give this implementation a shot, I'd be delighted :) Let me know if anyone has any questions.

CC @aight8

**Comment by aight8** [comment]
For the enter part: When you insert the new item you have to start w/o scale: 0 to calculate the items delta coordinates. After calculation of the positions the scale: 0 need's to be applied. I don't think you will see any kind of flicker since I the browser don't render this, but keep that in mind.

The rest seems very reasonable even this preset thing

**Comment by joshwcomeau** [comment]
Right, yeah. Thanks to `requestAnimationFrame`, we can access computed layout positions before the paint, so we can apply all this without flicker :) I remember having some weirdness with Safari (I think I had to let an extra frame pass?), but for the most part, things work as expected.

**Comment by joshwcomeau** [comment]
Just a heads-up that I've got a basic implementation working perfectly :) There were a bunch of implementation issues but I think I've solved most of the problems. Will spend some time this weekend figuring out what the public API ought to be, updating the docs, and adding some sort of demo.

So yeah, help no longer needed, but I'll link to a PR soon.

**Comment by maxsalven** [comment]
Fantastic, thank you! One idea for animations might be to have new items enter from eg above the browser window, and departing items exit via eg the bottom of the browser window. Flags control which specific direction things enter/exit from. The shrink/expand theme sounds great too.

**Comment by joshwcomeau** [comment]
Actually, I could really use feedback for the API.

Right now there's a single new prop, `enterExitAnimation`. You can pass it:

* `false` to disable enter/exit animations
* a String, to select one of the presets (right now I have "elevator", "fade", "accordianVertical", and "accordianHorizontal"
* an Object with the custom properties you want.

For the custom object, it expects something like:

```js
{
  enter: {
    from: {
      transform: "translate(-100px)",
      opacity: 0
    },
    to: {
      transform: '',
      opacity: 1
    }
  }, 
  leave: { /* same stuff */ }
}
```

The idea is that any valid CSS can be supplied, to be triggered on enter/leave start/finish.

It's the best compromise I've found between total control and convenience. I expect most people will never need to use the custom object, since Im going to make most common options available via presets, but it'll be good to have if you need it.

@maxsalven your example is neat, and it'd be possible with the custom object, but you do bring up a good point; there may be neat "special" effects that can be provided through the presets, instead of the presets just resolving to basic CSS.

@aight8 thoughts?

@marcus433 because I'm modifying styles, this might conflict with your work on preserving CSS transform property, but we should be able to use the same parsing-and-diffing trick :)

**Comment by marcus433** [comment]
I can definitely get this working with the transform preserver. Regarding the API, I think presets and a custom object are a great way to go about it.

**Comment by pvande** [comment]
@joshwcomeau Question: any reason not to break apart the prop into separate enter and leave variants?  `<Element enterTransition="fade" exitTransition={false} />` seems a little cleaner and more natural, without complicating the implementation too much.

I'm unsure whether `exit` or `leave` is more correct, or whether the properties would be better named `enterAnimation`, `enterEffect`, or `enterTransition`.

+1 on the "escape hatch" of having arbitrary styles as a fallback, though you may want to validate them  and warn if they aren't strictly FLIP-friendly.

**Comment by joshwcomeau** [comment]
@pvande Thanks for the feedback!

I made a mistake in my initial post; Related to exit vs. leave, I've decided to go with `leave` because it conforms more to convention. JS has `mouseenter` / `mouseleave`, ReactCSSTransitionGroup uses `transitionEnter` / `transitionLeave`, etc.

I'm pretty conflicted on whether they should be 2 attributes or 1. I agree that having them separate looks and feels cleaner, but I also feel like it's more of a pain to write it out every time, since I believe 95%+ of the time, it'll be the exact same property. Lots of `enterTransition="fade" leaveTransition="fade"`.

@marcus433 Awesome :D glad to hear it.

**Comment by marcus433** [comment]
How about a `transition` property and if it isn't present then it checks for transitionEnter & transitionLeave?

**Comment by pvande** [comment]
@joshwcomeau `leaveTransition` feels grammatically strange to me (hence the mention of `exit`) but it's not awful, and as you point out, there's prior art.  (We could also get into a discussion as to whether "enter" and "exit" or "leave" are even the right terms for set addition and removal, but I'm not going to push the issue any further.)

I agree that the values will often be the same, but 95+% seems a little high.  @marcus433's proposal of having a separate property for the common case of identical (preset!) transitions makes good sense if we can find a good name for it, and has the advantage of being a little easier to reason about programatically (e.g. fewer assertions about the schema of the passed value).

**Comment by joshwcomeau** [comment]
Yeah, I thought about having a default, but that's extra logic that people have to learn about. Doesn't feel as simple as two properties.

As it stands now, I have `enterAnimation` and `leaveAnimation`. I think I want to use "animation" over "transition" because to me, `enterTransition` feels like it should take a duration for the animation, not a description of the animation itself. Partially because that's how ReactCSSTransitionGroup uses it, but also because that's how the CSS property itself is most-often used.

I don't want to spend _too_ much time on the semantics, unless you guys think there's a significant cost to this choice. Naming things is important, but at the end of the day, as long as developers can easily figure out what it does, I'm happy.

Appreciate all the feedback :) I've been staring at this for a long time today.

**Comment by marcus433** [comment]
What i'm proposing is if they are the same one only needs an `animation` property. Of course it isn't necessary, they could still just use a `enterAnimation` and `leaveAnimation` but it is a nice thing to have. It shouldn't lead to people having to learn anything more as it is entirely optional.

Thus:

`<FlipMove animation="fade"></FlipMove>`
can still be represented as
`<FlipMove enterAnimation="fade" leaveAnimation="fade"></FlipMove>`

**Comment by joshwcomeau** [comment]
Yeah, I understand. I still think it could lead to some confusion; I always find it a bit odd when there are multiple ways to do the same thing, even if it's meant as a convenience.

The bigger issue I see is naming it. `animation` is too general, since the primary purpose of this library is to animate list re-orderings, and this doesn't affect that. `enterLeaveAnimation` was my original idea, but it can be confusing as well.

If you guys are interested, the source is here: https://github.com/joshwcomeau/react-flip-move/tree/feature/enterExitAnimations

I still have to finish updating the demo site, and I'm sure I'll want to refactor this in the morning, but it's working pretty damn well and I'm happy about that :)

**Comment by pvande** [comment]
I agree completely on the naming and semantics issues; I just wanted to make sure the questions were at least mentioned.  "xAnimation" is a perfectly sensible choice, and is likely to be easily discovered.

`changeAnimation` (or `onChangeAnimation`) could be a decent stand-in for the convenience name, but requires an understanding that it's watching for changes to the *set*, not the *ordered list*.  (`reorderAnimation` or `moveAnimation` would cover the latter case, but … well, that's not necessarily a customization you're likely to want to support.)  It also may not align super well with the `enter` and `leave` model.

Overall, I'm just happy to see the functionality implemented, and look forward to using it soon.  :)

**Comment by joshwcomeau** [comment]
Launched :D use version 2.0.2 for enter/leave animations.

Thanks everyone for your feedback. The API is much better as a result :)

Let me know if you run into any weirdness; I did a fair bit of testing and caught half a dozen obscure bugs, but there may be more. This was a significant rewrite, which nearly doubled the amount of code involved.
